src/impl/collection_change_builder.cpp

```diff
@@ -316,6 +321,98 @@ void CollectionChangeBuilder::move_over(size_t row_ndx, size_t last_row, bool tr
     verify();
 }
 
+void CollectionChangeBuilder::swap(size_t ndx_1, size_t ndx_2, bool track_moves)
+{
+    REALM_ASSERT(ndx_1 != ndx_2);
+    // The order of the two indices doesn't matter semantically, but making them
+    // consistent simplifies the logic
+    if (ndx_1 > ndx_2)
+        std::swap(ndx_1, ndx_2);
+
+    bool row_1_modified = modifications.contains(ndx_1);
+    bool row_2_modified = modifications.contains(ndx_2);
+    if (row_1_modified != row_2_modified) {
+        if (row_1_modified) {
+            modifications.remove(ndx_1);
+            modifications.add(ndx_2);
+        }
+        else {
+            modifications.remove(ndx_2);
+            modifications.add(ndx_1);
+        }
+    }
+
+    if (!track_moves)
+        return;
+
+    auto update_move = [&](auto existing_it, auto ndx_1, auto ndx_2) {
+        // update the existing move to ndx_2 to point at ndx_1
+        auto original = existing_it->second;
+        m_move_mapping.erase(existing_it);
+        m_move_mapping[ndx_1] = original;
+
+        // add a move from 1 -> 2 unless 1 was a new insertion
+        if (!insertions.contains(ndx_1)) {
+            m_move_mapping[ndx_2] = deletions.add_shifted(insertions.unshift(ndx_1));
+            insertions.add(ndx_1);
+        }
+        REALM_ASSERT_DEBUG(insertions.contains(ndx_2));
+    };
+
+    auto move_1 = m_move_mapping.find(ndx_1);
+    auto move_2 = m_move_mapping.find(ndx_2);
+    bool have_move_1 = move_1 != end(m_move_mapping) && move_1->first == ndx_1;
+    bool have_move_2 = move_2 != end(m_move_mapping) && move_2->first == ndx_2;
+    if (have_move_1 && have_move_2)
+        // both are already moves, so just swap the destinations
+        std::swap(move_1->second, move_2->second);
+    else if (have_move_1) {
```

Comment:
The absence of braces around the body of the preceding if is a little out of place, given the use of braces on the elses that follow.